File: taipei-streetlight/project/backend/src/main/java/com/taipei/iot/device/entity/Device.java
package com.taipei.iot.device.entity;

import com.taipei.iot.device.enums.ConnectivityType;
import com.taipei.iot.device.enums.DeviceStatus;
import com.taipei.iot.device.enums.DeviceType;
import com.taipei.iot.tenant.TenantAware;
import com.taipei.iot.tenant.TenantEntityListener;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EntityListeners;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.Filter;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.Map;

@Entity
@Table(name = "devices")
@Filter(name = "tenantFilter", condition = "tenant_id = :tenantId")
@EntityListeners({TenantEntityListener.class, AuditingEntityListener.class})
@Getter @Setter @Builder @NoArgsConstructor @AllArgsConstructor
public class Device implements TenantAware {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "tenant_id", nullable = false, length = 50)
    private String tenantId;

    @Enumerated(EnumType.STRING)
    @Column(name = "device_type", nullable = false, length = 30)
    private DeviceType deviceType;

    @Column(name = "device_code", nullable = false, length = 100)
    private String deviceCode;

    @Column(name = "device_name", length = 200)
    private String deviceName;

    // 坐標
    @Column(name = "twd97_x", precision = 12, scale = 3)
    private BigDecimal twd97X;

    @Column(name = "twd97_y", precision = 12, scale = 3)
    private BigDecimal twd97Y;

    @Column(name = "lng", precision = 11, scale = 7)
    private BigDecimal lng;

    @Column(name = "lat", precision = 10, scale = 7)
    private BigDecimal lat;

    @Column(name = "elevation", precision = 8, scale = 3)
    private BigDecimal elevation;

    @Column(name = "twd67_x", precision = 12, scale = 3)
    private BigDecimal twd67X;

    @Column(name = "twd67_y", precision = 12, scale = 3)
    private BigDecimal twd67Y;

    @Column(name = "taipower_coord", length = 100)
    private String taipowerCoord;

    // 組織歸屬
    @Column(name = "dept_id")
    private Long deptId;

    @Column(name = "contract_id")
    private Long contractId;

    @Column(name = "property_owner", length = 200)
    private String propertyOwner;

    // 狀態
    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false, length = 20)
    private DeviceStatus status;

    @Column(name = "installed_at")
    private LocalDate installedAt;

    @Column(name = "decommissioned_at")
    private LocalDate decommissionedAt;

    // 連線拓撲 / 物理組裝
    @Column(name = "parent_device_id")
    private Long parentDeviceId;

    @Column(name = "mount_position", length = 50)
    private String mountPosition;

    @Enumerated(EnumType.STRING)
    @Column(name = "connectivity_type", length = 20)
    private ConnectivityType connectivityType;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "network_config", columnDefinition = "jsonb")
    private Map<String, Object> networkConfig;

    @Column(name = "last_heartbeat_at")
    private LocalDateTime lastHeartbeatAt;

    // 回路
    @Column(name = "circuit_id")
    private Long circuitId;

    // ── IoT 擴充欄位 (Phase 7) ──
    @Column(name = "device_token", length = 200, unique = true)
    private String deviceToken;

    @Column(name = "auth_type", length = 20)
    private String authType;

    @Column(name = "firmware_version", length = 50)
    private String firmwareVersion;

    @Column(name = "last_telemetry_at")
    private LocalDateTime lastTelemetryAt;

    @Column(name = "format_id")
    private Long formatId;

    // 專有欄位
    @JdbcTypeCode(SqlTypes.JSON)
    @Column(name = "attributes", columnDefinition = "jsonb")
    private Map<String, Object> attributes;

    // 審計
    @Column(name = "created_by", length = 50)
    private String createdBy;

    @CreatedDate
    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @LastModifiedDate
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @Override
    public String getTenantId() { return tenantId; }

    @Override
    public void setTenantId(String tenantId) { this.tenantId = tenantId; }
}

File: taipei-streetlight/project/frontend/src/types/device.ts
export type DeviceType = 'POLE' | 'LUMINAIRE' | 'PANEL_BOX' | 'CONTROLLER' | 'POWER_EQUIPMENT' | 'ATTACHMENT'
export type DeviceStatus = 'ACTIVE' | 'REPORTED' | 'UNDER_REPAIR' | 'INACTIVE' | 'DECOMMISSIONED'
export type ConnectivityType = 'NONE' | 'DIRECT' | 'GATEWAY'

export interface DeviceResponse {
  id: number
  deviceType: DeviceType
  deviceCode: string
  deviceName: string | null
  twd97X: number | null
  twd97Y: number | null
  lng: number | null
  lat: number | null
  elevation: number | null
  twd67X: number | null
  twd67Y: number | null
  taipowerCoord: string | null
  deptId: number | null
  deptName: string | null
  contractId: number | null
  propertyOwner: string | null
  status: DeviceStatus
  installedAt: string | null
  decommissionedAt: string | null
  parentDeviceId: number | null
  parentDeviceCode: string | null
  mountPosition: string | null
  connectivityType: ConnectivityType | null
  networkConfig: Record<string, unknown> | null
  lastHeartbeatAt: string | null
  circuitId: number | null
  circuitNumber: string | null
  attributes: Record<string, unknown> | null
  childrenCount: number
  children: DeviceResponse[] | null
  createdBy: string | null
  createdAt: string
  updatedAt: string
}

export interface DeviceRequest {
  deviceType: DeviceType
  deviceCode: string
  deviceName?: string
  twd97X?: number
  twd97Y?: number
  lng?: number
  lat?: number
  elevation?: number
  twd67X?: number
  twd67Y?: number
  taipowerCoord?: string
  deptId?: number | null
  contractId?: number
  propertyOwner?: string
  installedAt?: string
  parentDeviceId?: number
  mountPosition?: string
  connectivityType?: ConnectivityType
  networkConfig?: Record<string, unknown>
  circuitId?: number
  attributes?: Record<string, unknown>
}

export interface ComponentReplaceRequest {
  oldDeviceId: number
  newDevice: DeviceRequest
  reason?: string
}

export interface PageResponse<T> {
  content: T[]
  totalElements: number
  totalPages: number
  page: number
  size: number
}
